File: 01-reference-agent/cpp/agent.cpp
#include "agent.hpp"

#include <nlohmann/json.hpp>

#include <cctype>
#include <chrono>
#include <cstdlib>
#include <curl/curl.h>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <thread>

using json = nlohmann::json;
namespace fs = std::filesystem;

namespace {

constexpr const char* kCannedInput =
    "No ticker stories loaded yet. Ask the user to click Get Stories, "
    "then produce a brief placeholder note that you are waiting for headlines.";
constexpr const char* kDefaultBedrockModelId = "us.amazon.nova-lite-v1:0";

std::string g_mode_override;

std::string trim_copy(std::string s) {
    while (!s.empty() && std::isspace(static_cast<unsigned char>(s.front()))) {
        s.erase(s.begin());
    }
    while (!s.empty() && std::isspace(static_cast<unsigned char>(s.back()))) {
        s.pop_back();
    }
    return s;
}

std::string lower_copy(std::string s) {
    for (char& ch : s) {
        ch = static_cast<char>(std::tolower(static_cast<unsigned char>(ch)));
    }
    return s;
}

std::string env_or(const char* key, const char* fallback) {
    if (const char* v = std::getenv(key)) {
        std::string s = trim_copy(v);
        if (!s.empty()) {
            return s;
        }
    }
    return fallback ? fallback : "";
}

std::string load_system_prompt() {
    const fs::path path = fs::path(example_root()) / "prompts" / "system_prompt.txt";
    std::ifstream in(path);
    if (!in) {
        throw std::runtime_error("could not read system prompt at " + path.string());
    }
    std::ostringstream ss;
    ss << in.rdbuf();
    std::string text = trim_copy(ss.str());
    if (text.empty()) {
        throw std::runtime_error("system prompt file is empty: " + path.string());
    }
    return text;
}

std::string build_user_input(const std::vector<TickerBlock>& ticker_results) {
    if (ticker_results.empty()) {
        return kCannedInput;
    }
    return format_stories_for_prompt(ticker_results);
}

int estimate_tokens(const std::string& text) {
    return std::max(1, static_cast<int>(text.size() / 4));
}

void fill_token_estimates(const std::string& completion, Metrics& m,
                          const std::string& user_input) {
    std::string sys;
    try {
        sys = load_system_prompt();
    } catch (...) {
    }
    m.prompt_tokens = estimate_tokens(sys + user_input);
    m.completion_tokens = estimate_tokens(completion);
    m.total_tokens = *m.prompt_tokens + *m.completion_tokens;
}

std::string stub_response(const Persona& persona,
                          const std::vector<TickerBlock>& ticker_results) {
    std::ostringstream b;
    b << "[stub / default-no-llm]\n";
    b << "Persona: " << persona.name << " (" << persona.profile << ")\n\n";
    b << "Headline briefing (stub):\n";
    if (ticker_results.empty()) {
        b << "- (no stories loaded — click Get Stories)\n";
    } else {
        for (const auto& block : ticker_results) {
            const std::string ticker = block.ticker.empty() ? "?" : block.ticker;
            b << "- " << ticker << ":\n";
            if (block.stories.empty()) {
                b << "  (no stories)\n";
            }
            for (const auto& s : block.stories) {
                b << "  • " << (s.title.empty() ? "(untitled)" : s.title) << "\n";
            }
        }
    }
    b << "\nAs a " << persona.profile
      << " analyst, this is boilerplate report text for UI testing. "
         "Switch AGENT_LLM_MODE to ollama or bedrock for a real model response.";
    return b.str();
}

size_t write_cb(char* ptr, size_t size, size_t nmemb, void* userdata) {
    auto* out = static_cast<std::string*>(userdata);
    out->append(ptr, size * nmemb);
    return size * nmemb;
}

struct StreamBuf {
    std::string buffer;
    std::function<void(const std::string&)> on_chunk;
    std::string error;
};

size_t stream_write_cb(char* ptr, size_t size, size_t nmemb, void* userdata) {
    auto* sb = static_cast<StreamBuf*>(userdata);
    sb->buffer.append(ptr, size * nmemb);
    size_t pos = 0;
    while (true) {
        const auto nl = sb->buffer.find('\n', pos);
        if (nl == std::string::npos) {
            break;
        }
        std::string line = sb->buffer.substr(pos, nl - pos);
        pos = nl + 1;
        while (!line.empty() && (line.back() == '\r' || line.back() == ' ')) {
            line.pop_back();
        }
        if (line.empty()) {
            continue;
        }
        try {
            const json data = json::parse(line);
            if (data.contains("error") && !data["error"].is_null()) {
                sb->error = data["error"].dump();
                continue;
            }
            if (data.contains("message") && data["message"].is_object()) {
                const std::string content = data["message"].value("content", "");
                if (!content.empty() && sb->on_chunk) {
                    sb->on_chunk(content);
                }
            }
        } catch (...) {
            // skip malformed line
        }
    }
    if (pos > 0) {
        sb->buffer.erase(0, pos);
    }
    return size * nmemb;
}

void ollama_stream(const std::string& model,
                   const std::vector<TickerBlock>& ticker_results,
                   const std::function<void(const std::string&)>& on_chunk) {
    const std::string sys = load_system_prompt();
    json body = {{"model", model},
                 {"stream", true},
                 {"messages",
                  json::array({{{"role", "system"}, {"content", sys}},
                               {{"role", "user"},
                                {"content", build_user_input(ticker_results)}}})}};
    const std::string payload = body.dump();
    const std::string host = ollama_host();
    const std::string url = host + "/api/chat";

    StreamBuf sb;
    sb.on_chunk = on_chunk;
    CURL* curl = curl_easy_init();
    if (!curl) {
        throw std::runtime_error("curl_easy_init failed");
    }
    struct curl_slist* headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/json");
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, payload.c_str());
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 120L);
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, stream_write_cb);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &sb);
    const CURLcode rc = curl_easy_perform(curl);
    long status = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &status);
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
    if (rc != CURLE_OK) {
        throw std::runtime_error("Ollama request failed (" + host + "): " +
                                 curl_easy_strerror(rc) +
                                 ". Is Ollama running, and is OLLAMA_MODEL pulled?");
    }
    if (status < 200 || status >= 300) {
        throw std::runtime_error("Ollama request failed (" + host + "): HTTP " +
                                 std::to_string(status) +
                                 ". Is Ollama running, and is OLLAMA_MODEL pulled?");
    }
    if (!sb.error.empty()) {
        throw std::runtime_error(sb.error);
    }
}

}  // namespace

void set_mode_override(const std::string& mode) {
    const std::string cleaned = lower_copy(trim_copy(mode));
    if (cleaned.empty()) {
        g_mode_override.clear();
        return;
    }
    if (cleaned == "stub" || cleaned == "ollama" || cleaned == "bedrock" ||
        cleaned == "anthropic") {
        g_mode_override = cleaned;
    } else {
        g_mode_override = "stub";
    }
}

std::string resolve_mode() {
    std::string mode = g_mode_override;
    if (mode.empty()) {
        mode = lower_copy(trim_copy(env_or("AGENT_LLM_MODE", "stub")));
    }
    if (mode == "stub" || mode == "ollama" || mode == "bedrock" ||
        mode == "anthropic") {
        return mode;
    }
    return "stub";
}

std::string model_label(const std::string& mode) {
    const std::string override_model = env_or("AGENT_LLM_MODEL", "");
    if (!override_model.empty()) {
        return override_model;
    }
    if (mode == "stub") {
        return "default-no-llm";
    }
    if (mode == "ollama") {
        return env_or("OLLAMA_MODEL", "llama3.2:3b");
    }
    if (mode == "bedrock") {
        return env_or("AGENT_BEDROCK_MODEL_ID", kDefaultBedrockModelId);
    }
    if (mode == "anthropic") {
        return env_or("ANTHROPIC_MODEL", "claude-3-haiku-20240307");
    }
    return "(unknown)";
}

std::string ollama_host() {
    std::string host = env_or("OLLAMA_HOST", "http://127.0.0.1:11434");
    while (!host.empty() && host.back() == '/') {
        host.pop_back();
    }
    return host;
}

bool probe_ollama(int timeout_ms) {
    const std::string url = ollama_host() + "/api/tags";
    std::string body;
    CURL* curl = curl_easy_init();
    if (!curl) {
        return false;
    }
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS, static_cast<long>(timeout_ms));
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_cb);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &body);
    const CURLcode rc = curl_easy_perform(curl);
    long status = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &status);
    curl_easy_cleanup(curl);
    return rc == CURLE_OK && status >= 200 && status < 300;
}

std::string ensure_llm_mode() {
    if (const char* v = std::getenv("AGENT_LLM_MODE")) {
        if (!trim_copy(v).empty()) {
            return resolve_mode();
        }
    }
    if (probe_ollama(600)) {
        setenv("AGENT_LLM_MODE", "ollama", 1);
    } else {
        setenv("AGENT_LLM_MODE", "stub", 1);
    }
    return resolve_mode();
}

void generate_stream(const Persona& persona,
                     const std::vector<TickerBlock>& ticker_results,
                     const std::function<void(const StreamEvent&)>& on_event) {
    const std::string mode = resolve_mode();
    const std::string model = model_label(mode);
    const std::string user_input = build_user_input(ticker_results);

    StreamEvent meta;
    meta.type = StreamEvent::Type::Meta;
    meta.provider = mode;
    meta.model = model;
    meta.mode = mode;
    meta.input = user_input;
    on_event(meta);

    const auto started = std::chrono::steady_clock::now();
    Metrics metrics;

    auto elapsed_ms = [&]() -> long long {
        return std::chrono::duration_cast<std::chrono::milliseconds>(
                   std::chrono::steady_clock::now() - started)
            .count();
    };

    try {
        if (mode == "stub") {
            const std::string text = stub_response(persona, ticker_results);
            bool first = true;
            for (size_t i = 0; i < text.size(); i += 12) {
                if (first) {
                    metrics.ttft_ms = elapsed_ms();
                    first = false;
                }
                StreamEvent ev;
                ev.type = StreamEvent::Type::Token;
                ev.text = text.substr(i, 12);
                on_event(ev);
                std::this_thread::sleep_for(std::chrono::milliseconds(20));
            }
            metrics.finish_reason = "stop";
            fill_token_estimates(text, metrics, user_input);
        } else if (mode == "ollama") {
            std::string parts;
            bool first = true;
            ollama_stream(model, ticker_results, [&](const std::string& chunk) {
                if (first) {
                    metrics.ttft_ms = elapsed_ms();
                    first = false;
                }
                parts += chunk;
                StreamEvent ev;
                ev.type = StreamEvent::Type::Token;
                ev.text = chunk;
                on_event(ev);
            });
            metrics.finish_reason = "stop";
            fill_token_estimates(parts, metrics, user_input);
        } else if (mode == "bedrock") {
            StreamEvent ev;
            ev.type = StreamEvent::Type::Error;
            ev.message =
                "Mode 'bedrock' is not wired in the C++ example yet. "
                "Use AGENT_LLM_MODE=stub or ollama here, or run the Python web app "
                "for Bedrock.";
            on_event(ev);
            metrics.finish_reason = "error";
        } else {
            StreamEvent ev;
            ev.type = StreamEvent::Type::Error;
            ev.message = "Mode '" + mode +
                         "' is configured but not implemented in this reference yet. "
                         "Use AGENT_LLM_MODE=stub or ollama.";
            on_event(ev);
            metrics.finish_reason = "error";
        }
    } catch (const std::exception& ex) {
        StreamEvent ev;
        ev.type = StreamEvent::Type::Error;
        ev.message = ex.what();
        on_event(ev);
        metrics.finish_reason = "error";
    }

    metrics.latency_ms = elapsed_ms();
    StreamEvent m_ev;
    m_ev.type = StreamEvent::Type::Metrics;
    m_ev.metrics = metrics;
    on_event(m_ev);
    StreamEvent done;
    done.type = StreamEvent::Type::Done;
    on_event(done);
}

File: 20-agent-config/21-agent-completion-config/go/agent.go
// agent.go — domain logic for 21-agent-completion-config[go] (no TUI code here).
//
// =============================================================================
// HOW TO READ THIS FILE
// =============================================================================
//
// Same product flow as 01-reference-agent, but at generate time LaunchDarkly
// AgentControl supplies **model**, **system** message, and **user** message.
//
//  1. Data          Personas (UI labels + LD context key/name)
//  2. LaunchDarkly  Init server SDK + AI SDK; CompletionConfig evaluation
//  3. Providers     Route by served provider/model (Ollama Custom, …)
//  4. Generation    generateStream() — evaluate config, then stream LLM tokens
//
// LaunchDarkly insertion point (read this first):
//
//	generateStream() → aiClient.CompletionConfig(...)
//	Docs: https://launchdarkly.com/docs/sdk/ai/go
//	Keywords: AgentControl · completion config · AI SDK · message variables
//
// Variables: the config user message includes {{ stories }}; we pass
// {"stories": <formatted headlines>} so LaunchDarkly substitutes at evaluate time.
//
// API note (Go AI SDK is pre-1.0): the module is a *separate* Go module,
// github.com/launchdarkly/go-server-sdk-ai (package ldai), distinct from the
// older ldai subpackage bundled inside go-server-sdk/v7. Only the separate
// module exposes CompletionConfig; the bundled one only has a generic Config.
// See 20-agent-config/21-agent-completion-config/go/README.md for details.
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/launchdarkly/go-sdk-common/v3/ldcontext"
	ld "github.com/launchdarkly/go-server-sdk-ai/ldai"
	"github.com/launchdarkly/go-server-sdk-ai/ldai/datamodel"
	ldclient "github.com/launchdarkly/go-server-sdk/v7"
)

// ---------------------------------------------------------------------------
// 1. Data — demo personas (also become the LD evaluation context)
// ---------------------------------------------------------------------------

const (
	cannedStories = "No ticker stories loaded yet. Ask the user to click Get Stories."

	// LaunchDarkly: ai-config key=equity-briefing-completion name="Equity briefing completion" mode=completion
	// https://app.launchdarkly.com/projects/lunch-marcoly/ai-configs/equity-briefing-completion
	defaultConfigKey       = "equity-briefing-completion"
	defaultOllamaModelName = "llama3.2:3b"
)

// persona is the selectable demo identity — also the LaunchDarkly evaluation context.
type persona struct {
	ID        string
	Name      string
	Profile   string
	Anonymous bool
}

var personas = []persona{
	{ID: "conservative-charlie", Name: "Conservative Charlie", Profile: "conservative"},
	{ID: "neutral-nancy", Name: "Neutral Nancy", Profile: "neutral"},
	{ID: "thoughtless-toby", Name: "Thoughtless Toby", Profile: "risk-taker"},
	// No name targeting — anonymous context falls through to baseline-analyst.
	{ID: "anonymous-amelia", Name: "Anonymous Amelia", Profile: "anonymous", Anonymous: true},
}

type metrics struct {
	LatencyMS        *int
	TTFTMS           *int
	PromptTokens     *int
	CompletionTokens *int
	TotalTokens      *int
	FinishReason     string
}

type streamEvent struct {
	Type         string
	Persona      persona
	Input        string
	Provider     string
	Model        string
	Mode         string
	ConfigKey    string
	VariationKey string
	Fallback     bool
	Text         string
	Message      string
	Metrics      metrics
}

func configKey() string {
	if v := strings.TrimSpace(os.Getenv("LD_AGENT_CONFIG_KEY")); v != "" {
		return v
	}
	return defaultConfigKey
}

func defaultOllamaModel() string {
	if v := strings.TrimSpace(os.Getenv("OLLAMA_MODEL")); v != "" {
		return v
	}
	return defaultOllamaModelName
}

func ollamaHost() string {
	host := strings.TrimSpace(os.Getenv("OLLAMA_HOST"))
	if host == "" {
		host = "http://127.0.0.1:11434"
	}
	return strings.TrimRight(host, "/")
}

func formatStories(tickerResults []tickerBlock) string {
	if len(tickerResults) == 0 {
		return cannedStories
	}
	return formatStoriesForPrompt(tickerResults)
}

// Single source of truth with REST provisioning prompts: ../rest/messages/*.
func baselineMessagesDir() string {
	return filepath.Join(exampleRoot(), "rest", "messages")
}

func readMessageFile(name string) (string, error) {
	path := filepath.Join(baselineMessagesDir(), name)
	raw, err := os.ReadFile(path)
	if err != nil {
		return "", fmt.Errorf("could not read baseline message file %s: %w", path, err)
	}
	return string(raw), nil
}

// baselineSystemPrompt is the in-code baseline system prompt (same text as
// rest/messages/baseline-system.txt).
func baselineSystemPrompt() string {
	text, err := readMessageFile("baseline-system.txt")
	if err != nil {
		return ""
	}
	return strings.TrimSpace(text)
}

// baselineUserTemplate is the user prompt template with {{ stories }}
// (rest/messages/baseline-user.txt).
func baselineUserTemplate() string {
	text, err := readMessageFile("baseline-user.txt")
	if err != nil {
		return ""
	}
	return strings.TrimSpace(text)
}

// renderBaselineUser fills {{ stories }} locally when using the code baseline fallback.
func renderBaselineUser(storiesText string) string {
	out := strings.ReplaceAll(baselineUserTemplate(), "{{ stories }}", storiesText)
	return strings.ReplaceAll(out, "{{stories}}", storiesText)
}

// baselineMessages are the chat messages for the in-code baseline-analyst fallback.
func baselineMessages(storiesText string) []datamodel.Message {
	return []datamodel.Message{
		{Role: datamodel.System, Content: baselineSystemPrompt()},
		{Role: datamodel.User, Content: renderBaselineUser(storiesText)},
	}
}

func userMessageText(messages []datamodel.Message) string {
	for _, m := range messages {
		if m.Role == datamodel.User {
			return m.Content
		}
	}
	return ""
}

// ---------------------------------------------------------------------------
// 2. LaunchDarkly — server SDK + AI SDK (AgentControl)
// ---------------------------------------------------------------------------

var (
	sdkClient *ldclient.LDClient
	aiClient  *ld.Client
)

// initLaunchDarkly initializes the shared LaunchDarkly clients once at process start.
//
// LaunchDarkly: server-side SDK + AI SDK for AgentControl completion configs.
// https://launchdarkly.com/docs/sdk/ai/go
func initLaunchDarkly() error {
	if aiClient != nil {
		return nil
	}

	sdkKey := strings.TrimSpace(os.Getenv("LD_SDK_KEY"))
	if sdkKey == "" {
		return fmt.Errorf(
			"LD_SDK_KEY is required. Export a server-side SDK key for the environment that targets %s",
			configKey(),
		)
	}

	client, err := ldclient.MakeClient(sdkKey, 5*time.Second)
	if err != nil {
		return fmt.Errorf("LaunchDarkly client failed to initialize within 5s: %w. "+
			"Check LD_SDK_KEY and network access to LaunchDarkly", err)
	}
	if !client.Initialized() {
		return fmt.Errorf("LaunchDarkly client failed to initialize within 5s. " +
			"Check LD_SDK_KEY and network access to LaunchDarkly")
	}

	aic, err := ld.NewClient(client)
	if err != nil {
		_ = client.Close()
		return fmt.Errorf("could not create LaunchDarkly AI client: %w", err)
	}

	sdkClient = client
	aiClient = aic
	return nil
}

// buildContext builds the LD evaluation context for this persona.
//
// Named personas: user key + name (name targeting matches Charlie/Nancy/Toby).
// Anonymous Amelia: fixed key, anonymous=true — not indexed as a known user;
// name rules do not match → fallthrough (baseline-analyst).
// https://launchdarkly.com/docs/sdk/features/anonymous
func buildContext(p persona) ldcontext.Context {
	builder := ldcontext.NewBuilder(p.ID).Name(p.Name)
	if p.Anonymous {
		builder.Anonymous(true)
	}
	return builder.Build()
}

// baselineCompletionDefault is the SDK default when the config key is missing /
// unreachable. Also documents the intended offline shape. When the config exists
// but is turned **off**, LaunchDarkly still returns the disabled variation
// (enabled=false) — see generateStream() for the app-level fallback.
// https://launchdarkly.com/docs/sdk/ai/go
func baselineCompletionDefault() ld.AICompletionConfigDefault {
	return ld.NewAICompletionConfigDefault().
		WithEnabled(true).
		WithModelName(defaultOllamaModel()).
		WithProviderName("Custom").
		WithMessage(baselineSystemPrompt(), datamodel.System).
		WithMessage(baselineUserTemplate(), datamodel.User)
}

// evaluateCompletion fetches model + messages from AgentControl (completion mode).
//
// LaunchDarkly capability: CompletionConfig evaluation with message variables.
// https://launchdarkly.com/docs/sdk/features/agentcontrol-config
func evaluateCompletion(p persona, storiesText string) ld.AICompletionConfig {
	variables := map[string]interface{}{"stories": storiesText}
	return aiClient.CompletionConfig(configKey(), buildContext(p), baselineCompletionDefault(), variables)
}

// resolveRuntime maps served provider/model to a local caller (ollama | bedrock).
//
// Custom / Ollama models from rest/create-model-config.sh use provider Custom
// and a model id like llama3.2:3b → call local Ollama.
func resolveRuntime(model, providerName string) (provider, resolvedModel string, err error) {
	pl := strings.ToLower(strings.TrimSpace(providerName))

	switch {
	case pl == "custom" || pl == "ollama" || strings.Contains(model, ":"):
		return "ollama", model, nil
	case pl == "bedrock" || strings.HasPrefix(model, "us.") || strings.HasPrefix(model, "amazon.") ||
		strings.HasPrefix(model, "anthropic.") || strings.HasPrefix(model, "meta."):
		return "bedrock", model, nil
	case model == "":
		return "", "", fmt.Errorf("AgentControl variation has no model name. " +
			"Check modelConfigKey on the served variation in LaunchDarkly")
	default:
		// Unknown provider: try Ollama with the model id (classroom default).
		return "ollama", model, nil
	}
}

// ---------------------------------------------------------------------------
// 3. Providers — call whatever model AgentControl named
// ---------------------------------------------------------------------------

func estimateTokens(text string) int {
	n := len(text) / 4
	if n < 1 {
		return 1
	}
	return n
}

func fillTokenEstimates(messages []datamodel.Message, completion string, m *metrics) {
	var prompt strings.Builder
	for _, msg := range messages {
		prompt.WriteString(msg.Content)
	}
	pt := estimateTokens(prompt.String())
	ct := estimateTokens(completion)
	total := pt + ct
	m.PromptTokens = &pt
	m.CompletionTokens = &ct
	m.TotalTokens = &total
}

func ollamaStream(model string, messages []datamodel.Message, out chan<- string) error {
	payload := map[string]any{
		"model":    model,
		"stream":   true,
		"messages": messages,
	}
	raw, err := json.Marshal(payload)
	if err != nil {
		return err
	}
	host := ollamaHost()
	client := &http.Client{Timeout: 120 * time.Second}
	res, err := client.Post(host+"/api/chat", "application/json", bytes.NewReader(raw))
	if err != nil {
		return fmt.Errorf("ollama request failed (%s, model=%s): %w. "+
			"Is Ollama running, and does the AgentControl model id match `ollama list`?", host, model, err)
	}
	defer res.Body.Close()
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return fmt.Errorf("ollama request failed (%s, model=%s): HTTP %d. "+
			"Is Ollama running, and does the AgentControl model id match `ollama list`?", host, model, res.StatusCode)
	}

	scanner := bufio.NewScanner(res.Body)
	// Ollama chunks can be larger than default buffer.
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var data map[string]any
		if err := json.Unmarshal([]byte(line), &data); err != nil {
			continue
		}
		if errMsg, ok := data["error"]; ok && errMsg != nil {
			return fmt.Errorf("%v", errMsg)
		}
		if msg, ok := data["message"].(map[string]any); ok {
			if content := asString(msg["content"]); content != "" {
				out <- content
			}
		}
		if done, ok := data["done"].(bool); ok && done {
			return nil
		}
	}
	if err := scanner.Err(); err != nil && err != io.EOF {
		return err
	}
	return nil
}

func generateOllama(model string, messages []datamodel.Message, started time.Time, m *metrics, out chan<- streamEvent) error {
	chunks := make(chan string, 16)
	errCh := make(chan error, 1)
	go func() {
		errCh <- ollamaStream(model, messages, chunks)
		close(chunks)
	}()

	var parts []string
	first := true
	for chunk := range chunks {
		if first {
			ttft := int(time.Since(started).Milliseconds())
			m.TTFTMS = &ttft
			first = false
		}
		parts = append(parts, chunk)
		out <- streamEvent{Type: "token", Text: chunk}
	}
	if err := <-errCh; err != nil {
		return err
	}
	m.FinishReason = "stop"
	fillTokenEstimates(messages, strings.Join(parts, ""), m)
	return nil
}

// ---------------------------------------------------------------------------
// 4. Generation
// ---------------------------------------------------------------------------

func intOr(v *int, def int) int {
	if v == nil {
		return def
	}
	return *v
}

// generateStream evaluates AgentControl, then streams tokens from the served model.
//
// Event contract matches 01-reference-agent (meta / token / error / metrics / done),
// plus a "status" event used for the fallback notice.
//
// When the AgentControl config is **disabled**, unreachable, or the served
// variation carries no variation key (LD unreachable / config key missing —
// the Go AI SDK silently returns the caller-supplied default in that case),
// fall back to the in-code baseline-analyst prompts + local Ollama model —
// same text as rest/messages/baseline-*.txt.
func generateStream(p persona, tickerResults []tickerBlock) <-chan streamEvent {
	out := make(chan streamEvent, 32)
	go func() {
		defer close(out)
		storiesText := formatStories(tickerResults)
		started := time.Now()
		m := metrics{}

		emitDone := func() {
			lat := int(time.Since(started).Milliseconds())
			m.LatencyMS = &lat
			out <- streamEvent{Type: "metrics", Metrics: m}
			out <- streamEvent{Type: "done"}
		}

		usingFallback := false
		var config ld.AICompletionConfig
		var fallbackReason string

		if err := initLaunchDarkly(); err != nil {
			usingFallback = true
			fallbackReason = fmt.Sprintf("LaunchDarkly evaluation failed (%s); using code baseline.", err)
		} else {
			// LaunchDarkly: evaluate completion config (model + messages).
			config = evaluateCompletion(p, storiesText)
			if !config.Enabled() || config.VariationKey() == "" {
				usingFallback = true
				fallbackReason = fmt.Sprintf(
					"AgentControl config '%s' is off / enabled=false; using code baseline-analyst.", configKey())
			}
		}

		if usingFallback {
			messages := baselineMessages(storiesText)
			provider, model, mode := "ollama", defaultOllamaModel(), "baseline-fallback"
			fmt.Printf("[generate] %s: variation='code-baseline' reason='FALLBACK'\n", p.Name)

			promptPreview := userMessageText(messages)
			if promptPreview == "" {
				promptPreview = storiesText
			}
			out <- streamEvent{
				Type: "meta", Persona: p, Input: promptPreview,
				Provider: provider, Model: model + " (code baseline)", Mode: mode,
				ConfigKey: configKey(), Fallback: true,
			}
			if fallbackReason != "" {
				out <- streamEvent{Type: "status", Message: fallbackReason}
			}
			if err := generateOllama(model, messages, started, &m, out); err != nil {
				out <- streamEvent{Type: "error", Message: err.Error()}
				m.FinishReason = "error"
			}
			emitDone()
			return
		}

		provider, model, err := resolveRuntime(config.Model().Name, config.Provider().Name)
		messages := config.Messages()
		tracker := config.CreateTracker()
		if err == nil && len(messages) == 0 {
			err = fmt.Errorf("served variation has no messages")
		}
		if err != nil {
			out <- streamEvent{
				Type: "meta", Persona: p, Input: storiesText,
				Provider: "—", Model: "—", Mode: "launchdarkly", ConfigKey: configKey(),
			}
			out <- streamEvent{Type: "error", Message: err.Error()}
			m.FinishReason = "error"
			emitDone()
			return
		}

		fmt.Printf("[generate] %s: variation=%q\n", p.Name, config.VariationKey())
		promptPreview := userMessageText(messages)
		if promptPreview == "" {
			promptPreview = storiesText
		}
		out <- streamEvent{
			Type: "meta", Persona: p, Input: promptPreview,
			Provider: provider, Model: model, Mode: "launchdarkly",
			ConfigKey: configKey(), VariationKey: config.VariationKey(), Fallback: false,
		}

		switch provider {
		case "ollama":
			if err := generateOllama(model, messages, started, &m, out); err != nil {
				out <- streamEvent{Type: "error", Message: err.Error()}
				m.FinishReason = "error"
			}
		case "bedrock":
			out <- streamEvent{Type: "error", Message: "Bedrock is not wired in the Go example. " +
				"Retarget the variation to an Ollama / Custom model, or run the Python/Node/.NET web app for Bedrock."}
			m.FinishReason = "error"
		default:
			out <- streamEvent{Type: "error", Message: fmt.Sprintf("Unsupported runtime provider '%s'.", provider)}
			m.FinishReason = "error"
		}

		lat := int(time.Since(started).Milliseconds())
		m.LatencyMS = &lat

		// LaunchDarkly: report generation metrics for this run (tokens, latency, success/error).
		// https://launchdarkly.com/docs/sdk/features/ai-metrics
		if tracker != nil {
			if m.FinishReason == "error" {
				_ = tracker.TrackError()
			} else {
				_ = tracker.TrackSuccess()
				_ = tracker.TrackDuration(time.Duration(lat) * time.Millisecond)
				if m.TTFTMS != nil {
					_ = tracker.TrackTimeToFirstToken(time.Duration(*m.TTFTMS) * time.Millisecond)
				}
				if m.TotalTokens != nil || m.PromptTokens != nil || m.CompletionTokens != nil {
					_ = tracker.TrackTokens(ld.TokenUsage{
						Total:  intOr(m.TotalTokens, 0),
						Input:  intOr(m.PromptTokens, 0),
						Output: intOr(m.CompletionTokens, 0),
					})
				}
			}
		}

		out <- streamEvent{Type: "metrics", Metrics: m}
		out <- streamEvent{Type: "done"}
	}()
	return out
}
